lib/cc/analyzer/source_extractor.rb

```diff
@@ -0,0 +1,66 @@
+module CC
+  module Analyzer
+    class SourceExtractor
+      def initialize(source)
+        @source = source
+      end
+
+      def extract(location)
+        case
+        when (lines = location["lines"])
+          extract_from_lines(lines)
+        when (positions = location["positions"])
+          extract_from_positions(positions)
+        end
+      end
+
+      private
+
+      attr_reader :location, :source
+
+      def extract_from_lines(lines)
+        range = (lines["begin"] - 1..lines["end"] - 1)
+
+        source.each_line.with_object("").with_index do |(line, memo), index|
+          memo << line if range.include?(index)
+        end
+      end
+
+      def extract_from_positions(positions)
+        positions = convert_to_offsets(positions)
+
+        length = positions["end"]["offset"] - positions["begin"]["offset"]
+
+        source[positions["begin"]["offset"], length + 1]
+      end
+
+      def convert_to_offsets(positions)
+        positions.each do |key, value|
+          unless value["offset"]
+            positions[key] = {
+              "offset" => to_offset(value["line"] - 1, value["column"] - 1),
+            }
+          end
+        end
+
+        positions
+      end
+
+      def to_offset(line, column)
+        offset = 0
+
+        source.each_line.with_index do |source_line, index|
+          if line == index
+            offset += column
+          else
+            offset += source_line.length
+          end
```

Comment:
FWIW: I've started to come around to:

```rb
offset +=
  if line == index
    column
  else
    source_line.length
  end
```

Which appeases rubocop without misaligning the if/else/end depending on the size of the LHS of the assignment.